lib/mirage.ml

```diff
@@ -1890,60 +1910,39 @@ let link info name target target_debug =
-               of_list c_artifacts %% of_list static_libs % "-o" % out
-               %% of_list ldpostflags)
-    in
-    Log.info (fun m -> m "linking with %a" Bos.Cmd.pp linker);
-    Bos.OS.Cmd.run linker >>= fun () ->
-    Ok out
-  | `Muen ->
+  | `Virtio | `Muen | `Ukvm ->
+    let pkg, post = solo5_pkg target in
     extra_c_artifacts "freestanding" libs >>= fun c_artifacts ->
     static_libs "mirage-solo5" >>= fun static_libs ->
-    ldflags "solo5-kernel-muen" >>= fun ldflags ->
-    ldpostflags "solo5-kernel-muen" >>= fun ldpostflags ->
-    let out = name ^ ".muen" in
+    ldflags pkg >>= fun ldflags ->
+    ldpostflags pkg >>= fun ldpostflags ->
+    let out = name ^ post in
+    let ld = find_ld pkg in
     let linker =
-      Bos.Cmd.(v "ld" %% of_list ldflags % "_build/main.native.o" %%
+      Bos.Cmd.(v ld %% of_list ldflags % "_build/main.native.o" %%
                of_list c_artifacts %% of_list static_libs % "-o" % out
                %% of_list ldpostflags)
     in
     Log.info (fun m -> m "linking with %a" Bos.Cmd.pp linker);
     Bos.OS.Cmd.run linker >>= fun () ->
-    Ok out
-  | `Ukvm ->
-    extra_c_artifacts "freestanding" libs >>= fun c_artifacts ->
-    static_libs "mirage-solo5" >>= fun static_libs ->
-    ldflags "solo5-kernel-ukvm" >>= fun ldflags ->
-    ldpostflags "solo5-kernel-ukvm" >>= fun ldpostflags ->
-    let out = name ^ ".ukvm" in
-    let linker =
-      Bos.Cmd.(v "ld" %% of_list ldflags % "_build/main.native.o" %%
-               of_list c_artifacts %% of_list static_libs % "-o" % out
-               %% of_list ldpostflags)
-    in
-    let ukvm_mods =
-      List.fold_left (fun acc -> function
-        | "mirage-net-solo5" -> "net" :: acc
-        | "mirage-block-solo5" -> "blk" :: acc
-        | _ -> acc)
-        [] libs @ (if target_debug then ["gdb"] else [])
-    in
-    pkg_config "solo5-kernel-ukvm" ["--variable=libdir"] >>= function
-    | [ libdir ] ->
-      Bos.OS.Cmd.run Bos.Cmd.(v "ukvm-configure" % (libdir ^ "/src/ukvm") %% of_list ukvm_mods) >>= fun () ->
-      Bos.OS.Cmd.run Bos.Cmd.(v "make" % "-f" % "Makefile.ukvm" % "ukvm-bin") >>= fun () ->
-      Log.info (fun m -> m "linking with %a" Bos.Cmd.pp linker);
-      Bos.OS.Cmd.run linker >>= fun () ->
+    if target = `Ukvm then
+      let ukvm_mods =
+        List.fold_left (fun acc -> function
+            | "mirage-net-solo5" -> "net" :: acc
+            | "mirage-block-solo5" -> "blk" :: acc
+            | _ -> acc)
+          [] libs @ (if target_debug then ["gdb"] else [])
+      in
+      pkg_config pkg ["--variable=libdir"] >>= function
+      | [ libdir ] ->
+        Bos.OS.Cmd.run Bos.Cmd.(v "ukvm-configure" % (libdir ^ "/src/ukvm") %% of_list ukvm_mods) >>= fun () ->
+        Bos.OS.Cmd.run Bos.Cmd.(v "make" % "-f" % "Makefile.ukvm" % "ukvm-bin") >>= fun () ->
+        Log.info (fun m -> m "linking with %a" Bos.Cmd.pp linker);
+        Bos.OS.Cmd.run linker >>= fun () ->
```

Comment:
@hannesm Maybe I'm misreading the diff, but won't this run the linker twice for `ukvm`? The first call is on line 1927 just before the `if target = Ukvm`.

Reply:
well spotted, I pushed 093ba78d which removes the second call to linker